File: examples/v1/workflows/on-failure.go
package v1_workflows

import (
	"errors"

	v1 "github.com/hatchet-dev/hatchet/pkg/v1"
	"github.com/hatchet-dev/hatchet/pkg/v1/task"
	"github.com/hatchet-dev/hatchet/pkg/v1/workflow"
	"github.com/hatchet-dev/hatchet/pkg/worker"
)

type AlwaysFailsOutput struct {
	TransformedMessage string
}

type OnFailureOutput struct {
	FailureRan bool
}

type OnFailureSuccessResult struct {
	AlwaysFails AlwaysFailsOutput
}

func OnFailure(hatchet *v1.HatchetClient) workflow.WorkflowDeclaration[any, OnFailureSuccessResult] {

	simple := v1.WorkflowFactory[any, OnFailureSuccessResult](
		workflow.CreateOpts[any]{
			Name: "on-failure",
			OnFailureTask: &task.OnFailureTaskDeclaration[any]{
				Fn: func(_ any, ctx worker.HatchetContext) (*OnFailureOutput, error) {
					return &OnFailureOutput{
						FailureRan: true,
					}, nil
				},
			},
		},
		hatchet,
	)

	simple.Task(
		task.CreateOpts[any]{
			Name: "AlwaysFails",
			Fn: func(_ any, ctx worker.HatchetContext) (*AlwaysFailsOutput, error) {
				return &AlwaysFailsOutput{
					TransformedMessage: "always fails",
				}, errors.New("always fails")
			},
		},
	)

	return simple
}

File: sdks/typescript/src/examples/on-failure.ts
import Hatchet from '../sdk';
import { Workflow } from '../workflow';

const hatchet = Hatchet.init();

// ❓ OnFailure Step
// This workflow will fail because the step will throw an error
// we define an onFailure step to handle this case

const workflow: Workflow = {
  // ... normal workflow definition
  id: 'on-failure-example',
  description: 'test',
  on: {
    event: 'user:create',
  },
  // ,
  steps: [
    {
      name: 'step1',
      run: async (ctx) => {
        // 👀 this step will always throw an error
        throw new Error('Step 1 failed');
      },
    },
  ],
  // 👀 After the workflow fails, this special step will run
  onFailure: {
    name: 'on-failure-step',
    run: async (ctx) => {
      // 👀 we can do things like perform cleanup logic
      // or notify a user here

      // 👀 you can access the error from the failed step(s) like this
      console.log(ctx.errors());

      return { onFailure: 'step' };
    },
  },
};
// ‼️

// ❓ OnFailure With Details
// Coming soon to TypeScript! https://github.com/hatchet-dev/hatchet-typescript/issues/447
// ‼️

async function main() {
  const worker = await hatchet.worker('example-worker', 1);
  await worker.registerWorkflow(workflow);
  worker.start();
}

main();
